### PR TITLE
[feature] allow creating "rich domain models"

### PR DESCRIPTION
**Pull request by kbond**
Possible solution for #305. When forcing properties with the `Instantiator`, covert `array`'s to `ArrayCollection`'s if the typehint demands it.

Not a huge fan of this solution... I think I'd prefer `FactoryCollection` generate an `ArrayCollection` and then it can be set directly. I'll explore this option.

### CODE REVIEW

File: tests/Fixtures/Entity/RichPost.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * @ORM\Column(type="string", length=255)
+     */
+    private $title;
+
+    /**
+     * @ORM\OneToMany(targetEntity=RichComment::class, mappedBy="post")
+     */
+    private Collection $comments;
```
Comment (kbond) [inline tests/Fixtures/Entity/RichPost.php:29]: This typehint causes problems when running tests on < PHP 7.4...

File: tests/Fixtures/Factories/RichPostFactory.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+use Zenstruck\Foundry\Instantiator;
+use Zenstruck\Foundry\ModelFactory;
+use Zenstruck\Foundry\Tests\Fixtures\Entity\RichPost;
+
+final class RichPostFactory extends ModelFactory
+{
+    protected function initialize()
+    {
+        return $this->instantiateWith((new Instantiator())->alwaysForceProperties());
```
Comment (nikophil) [inline tests/Fixtures/Factories/RichPostFactory.php]: nice, I didn't know this one. This way no need for any setter :tada:

File: tests/Fixtures/Entity/RichPost.php (hunk 1 above)
Comment (nikophil) [reply tests/Fixtures/Entity/RichPost.php:29]: I think we can skip the related test for lower versions

File: tests/Functional/ORMModelFactoryTest.php [hunk 3]
```diff
@@ -493,6 +495,24 @@ public function embeddables_are_never_persisted(): void
...
+        $post = RichPostFactory::createOne([
+            'comments' => RichCommentFactory::new()->many(5),
+        ]);
+
+        RichPostFactory::assert()->count(1);
+        RichCommentFactory::assert()->count(5);
+        $this->assertCount(5, $post->getComments());
+
+        foreach (RichCommentFactory::all() as $comment) {
+            $this->assertSame($post->getId(), $comment->getPost()->getId());
```
Comment (nikophil) [inline tests/Functional/ORMModelFactoryTest.php]: I don't understand how this works :upside_down_face: 
we never say "the post we're creating is the one to use for all comments". How is it possible that a new post is not created for each comment?

File: src/Instantiator.php [hunk 4]
```diff
@@ -130,6 +132,19 @@ public function alwaysForceProperties(array $properties = []): self
      */
     public static function forceSet(object $object, string $property, $value): void
     {
+        $refProp = self::accessibleProperty($object, $property);
+
+        // if value is an array and the property expects a doctrine collection, convert to doctrine collection
+        if (\is_array($value) && \method_exists($refProp, 'getType') && $type = $refProp->getType()) {
```
Comment (nikophil) [inline src/Instantiator.php]: indeed, directly returning array collection would really be a better fit, and seems simpler (deprecation layer excepted...)

File: tests/Fixtures/Entity/RichPost.php (hunk 1 above)
Comment (kbond) [reply tests/Fixtures/Entity/RichPost.php:29]: It's not that easy as the `RichPost` will be loaded on every version.

File: tests/Fixtures/Factories/RichPostFactory.php (hunk 2 above)
Comment (kbond) [reply tests/Fixtures/Factories/RichPostFactory.php]: You can set this globally in your config too.

File: tests/Functional/ORMModelFactoryTest.php (hunk 3 above)
Comment (kbond) [reply tests/Functional/ORMModelFactoryTest.php]: There is some magic in `Factory` that handles this when creating relationships.

Comment (nikophil) [reply tests/Functional/ORMModelFactoryTest.php]: nice! it was this particular magic I didn't know about when I wrote the issue

File: tests/Fixtures/Entity/RichPost.php (hunk 1 above)
Comment (nikophil) [reply tests/Fixtures/Entity/RichPost.php:29]: I got another idea :joy: 
https://www.php.net/supported-versions.php

Comment (nikophil) [reply tests/Fixtures/Entity/RichPost.php:29]: I got another idea :imp: 
Capture d’écran du 2022-10-13 17-40-32

Comment (kbond) [reply tests/Fixtures/Entity/RichPost.php:29]: I know. I think we should bump the min PHP version to 8.0 in `1.x` and leave a `1.?.x` branch for bug fixes.

Comment (nikophil) [reply tests/Fixtures/Entity/RichPost.php:29]: I'm a lot in favor for this move :+1:

File: tests/Functional/ORMModelFactoryTest.php (hunk 3 above)
Comment (Kocal) [inline tests/Functional/ORMModelFactoryTest.php]: What about asserting between objects instead of scalars?
```suggestion
            $this->assertSame($post, $comment->getPost());
```

File: tests/Fixtures/Entity/RichComment.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    private $post;
+
+    public function __construct(string $body, RichPost $post)
+    {
+        $this->body = $body;
+        $this->post = $post;
+    }
+
+    public function getId(): ?int
```
Comment (benblub) [inline tests/Fixtures/Entity/RichComment.php:36]: getId() return type should be `int`

Comment (kbond) [reply tests/Fixtures/Entity/RichComment.php:36]: What about when the id truly is null (before the entity is persisted)?

Comment (benblub) [reply tests/Fixtures/Entity/RichComment.php:36]: good point.

Comment (nikophil) [reply tests/Fixtures/Entity/RichComment.php:36]: if this entity is part of a "rich model" it should use uuids and set the uuid in its constructor :innocent: 

but I think we don't really care for these fixtures :shrug:

Comment (kbond) [reply tests/Fixtures/Entity/RichComment.php:36]: This is very true! (I didn't want to add a dep)

**Comment by X-Coder264** [comment]
@kbond Is there anything left to be done before this can be merged?

We are currently trying to migrate to Foundry but this is blocking us.

As CommentFactory in our case has `'post' => PostFactory::new(),` in its `getDefaults` method (so that the `CommentFactory::createOne();` call can work) the piece of code below creates two post records and one comment record in the database (instead of the expected one post record with one comment). Plus the created comment record gets created for the other post, not the one that actually gets returned by the post factory call.

```php
        $post = PostFactory::createOne([
            'comments' => CommentFactory::new()->many(1),
        ])->object();

       $post->getComments()->isEmpty(); // currently true, the expected result is false
```

**Comment by nikophil** [comment]
Hi @X-Coder264 

> Is there anything left to be done before this can be merged?

we would like to try another way of doing this: `FactoryCollection::create()` could return directly an `ArrayCollection`. We could give a try in the next few days if this feature is missing for someone

**Comment by kbond** [comment]
Hey @X-Coder264, I'd like finish this one up shortly, just getting back from holidays.

**Comment by kbond** [comment]
I think this is ready as-is. @X-Coder264, can you try this out in your project?

There's some things I'd still like to do:
1. Split `Instantiator` into separate instantiator/hydrator (#388)
2. Use `symfony/property-info` to detect doctrine collections (right now the prop needs to be type-hinted as `Collection`).
3. Add `FactoryCollection::asDoctrineCollection()` or something like that.

But these can be added in the future (probably when we split the instantiator).

**Comment by kbond** [comment]
@nikophil, any idea on the CI failure? Related to #393 somehow?

**Comment by nikophil** [comment]
very strange :thinking:
I already had these errors in #393 but they were fixed by adding a condition on the migrations related to PHP 8.1
https://github.com/zenstruck/foundry/blob/1.x/tests/Fixtures/Migrations/Version20230109134337.php#L19

**Comment by nikophil** [comment]
would you mind rebasing on 1.x?

**Comment by kbond** [comment]
> would you mind rebasing on 1.x?

Done but still a problem. I see the failures from #393 that had the same issue, was [this](https://github.com/zenstruck/foundry/pull/393/files#diff-9f2b7d5a434ac908dce450d97972203bbf496cab5cd60547ffac77b374559a4f) the solution for that one?

**Comment by X-Coder264** [comment]
@kbond I've tried it out and for some reason I'm still getting the same old/wrong behavior even though I can see that the new `$value = new ArrayCollection($value);` code path gets successfully triggered when I put a Xdebug breakpoint there. I can't see any meaningful difference between your test case and my code (except that I use cascade persist, orphan removal and fetch extra lazy on my mapping but none of those should make a difference) :thinking: 

EDIT: It does make a difference.

This works (same scenario as your test case):

```php
    /**
     * @ORM\OneToMany(targetEntity="Comment", mappedBy="post")
     */
    private Collection $comments;
```

```php
    /**
     * @ORM\ManyToOne(targetEntity="Post", inversedBy="comments")
     * @ORM\JoinColumn(nullable=false)
     */
    private Post $post;
```

This doesn't work as expected:

```php
    /**
     * @ORM\OneToMany(targetEntity="Comment", mappedBy="post", cascade={"persist"}, orphanRemoval=true, fetch="EXTRA_LAZY")
     */
    private Collection $comments;
```

```php
    /**
     * @ORM\ManyToOne(targetEntity="Post", inversedBy="comments", cascade={"persist"})
     * @ORM\JoinColumn(nullable=false)
     */
    private Post $post;
```

**Comment by X-Coder264** [comment]
Ok, found the issue. Removing the cascade persist "fixes" the issue. So if I change my mapping to this it works as expected:

```php
    /**
     * @ORM\OneToMany(targetEntity="Comment", mappedBy="post", orphanRemoval=true, fetch="EXTRA_LAZY")
     */
    private Collection $comments;
```

I'm not sure why does cascade persist change anything, the behavior with or without cascade persist should be the same (getting back only one post with X comments).

**Comment by kbond** [comment]
Ok, thanks for testing and finding this edge case. I'll create a test case that uses cascade: persist and see if I can figure out why it doesn't work as expected.

There are some nuances with cascade persist that I don't fully understand (as I never use this feature myself). These are accounted for elsewhere in foundry but I didn't write this code. Maybe something with one of these is causing the problem.

**Comment by kbond** [comment]
I found the issue but not sure yet how to solve: 

When _NOT_ using cascade persist, the instantiation of the _comment_ collection is done after persisting the _post_. The _comments_ are instantiated with the previously created _post_ (this overrides the default post so the extra post per comment aren't created).

When using cascade persist, the collection needs to be instantiated before creating the post so I don't have a chance to set the post (as it isn't instantiated yet).

File: tests/Functional/ORMModelFactoryTest.php [hunk 6]
```diff
@@ -540,6 +545,102 @@ public function can_count_based_on_a_relationship(): void
         self::assertSame(2, PostFactory::count(['category' => $category]));
     }
 
+    /**
+     * @test
+     */
+    public function can_create_factories_for_rich_domain_models(): void
```
Comment (kbond) [inline tests/Functional/ORMModelFactoryTest.php:551]: This test passes and adds the comments via `RichPost::setComment()` (property-accessor).

File: tests/Functional/ORMModelFactoryTest.php [hunk 7]
```diff
@@ -540,6 +545,102 @@ public function can_count_based_on_a_relationship(): void
...
+        $post->save();
+
+        RichPostFactory::assert()->count(1);
+        RichCommentFactory::assert()->count(4);
+    }
+
+    /**
+     * @test
+     */
+    public function can_create_factories_for_rich_domain_models_using_force(): void
```
Comment (kbond) [inline tests/Functional/ORMModelFactoryTest.php:575]: This test passes and adds the comments by forcing `RichPost::$comments`.

File: tests/Functional/ORMModelFactoryTest.php [hunk 8]
```diff
@@ -540,6 +545,102 @@ public function can_count_based_on_a_relationship(): void
...
+        $post->save();
+
+        RichPostFactory::assert()->count(1);
+        RichCommentFactory::assert()->count(4);
+    }
+
+    /**
+     * @test
+     */
+    public function can_create_factories_for_cascade_rich_domain_models(): void
```
Comment (kbond) [inline tests/Functional/ORMModelFactoryTest.php:599]: Fails due to the cascade persist on `CascadePost::$comments`.

File: tests/Functional/ORMModelFactoryTest.php [hunk 9]
```diff
@@ -540,6 +545,102 @@ public function can_count_based_on_a_relationship(): void
...
+        $post->save();
+
+        CascadeRichPostFactory::assert()->count(1);
+        CascadeRichCommentFactory::assert()->count(4);
+    }
+
+    /**
+     * @test
+     */
+    public function can_create_factories_for_cascade_rich_domain_models_using_force(): void
```
Comment (kbond) [inline tests/Functional/ORMModelFactoryTest.php:623]: Fails due to the cascade persist on `CascadePost::$comments`.

**Comment by kbond** [comment]
I now test using `rich entities` with and without `alwaysForceProperties()` and this all works as expected.

The same tests using _cascade persist_ do not pass. Still not sure how to solve. We're kind of in a chicken-egg situation: to create the comment, we need the post but to create the post, we need the comment...

**Comment by kbond** [comment]
I think when we [split](https://github.com/zenstruck/foundry/issues/388) the instantiator into an instantiator and a hydrator we might be able to make `cascade: persist` work here.

> We're kind of in a chicken-egg situation: to create the comment, we need the post but to create the post, we need the comment...

We could _instantiate_ the post (w/o hydrating), then _instantiate_ and _hydrate_ the comments (using the post object), then hydrate the post with the created comments. Complicated but doable I think.

**Comment by nikophil** [comment]
I think we really should have this work with `cascade={"persist"}` since it is really common to have this

**Comment by kbond** [comment]
> I think we really should have this work with cascade={"persist"} since it is really common to have this

Agreed, I'm hoping #395 will help solve.

**Comment by nikophil** [comment]
@kbond I've found why some tests not related to this PR are failing. Could you add this code to all migration which does not have it?
```php
    public function isTransactional(): bool
    {
        return false;
    }
```
Migrations are only tested without DAMA, then they are only tested for "php8.0/sf4.4/lowest" and "php8.2/sf6.2/highest" but I don't understand why it does not fail in the second scenario... maybe `doctrine/migration` is clever enough to know transaction mode should be disabled in certain cases?

I'm considering to add a custom PHPStan rule since it's been multiple times I've lost time on this.

**Comment by nikophil** [comment]
hey @kbond 
could we close this PR now, since we'e merged https://github.com/zenstruck/foundry/pull/645 ?
